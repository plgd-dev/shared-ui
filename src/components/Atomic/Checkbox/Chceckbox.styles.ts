import { css } from '@emotion/react'
import styled from '@emotion/styled'
import { get, ThemeType } from '../_theme'

export const checkbox = css`
    display: flex;
    align-items: flex-start;
    position: relative;
    overflow: hidden;
`

export const check = styled.span`
    width: 16px;
    height: 16px;
    flex: 0 0 16px;
    border: 1px solid ${(props) => get(props.theme, `Checkbox.borderColor`)};
    border-radius: 4px;
    cursor: pointer;
    transition: all 0.25s;
    position: relative;
    box-sizing: border-box;
    top: 2px; // 22px LH

    &:before,
    &:after {
        content: '';
        display: block;
        transition: opacity 0.25s;
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
        width: 10px;
        border-radius: 2px;
        background: ${(props) => get(props.theme, `Checkbox.background`)};
        opacity: 0;
    }

    &:before {
        height: 2px;
    }

    &:after {
        height: 10px;
    }
`

export const input = (theme: ThemeType) => css`
    width: 1px;
    height: 1px;
    position: absolute;
    left: -10px;

    &:indeterminate ~ ${check} {
        border-color: ${get(theme, `Checkbox.input.borderColor`)};

        &:before {
            opacity: 1;
        }
    }

    &:checked ~ ${check} {
        border-color: ${get(theme, `Checkbox.input.borderColor`)};

        &:after {
            opacity: 1;
        }
    }
`

export const label = (theme: ThemeType) => css`
    color: ${get(theme, `Checkbox.label.color`)};
    padding-left: 8px;
    cursor: pointer;
`

export const error = (theme: ThemeType) => css`
    border-color: ${get(theme, `Checkbox.error.red`)};
`
