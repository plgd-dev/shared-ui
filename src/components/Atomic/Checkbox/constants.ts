export const checkboxValues = {
    ON: 'on',
    OFF: 'off',
} as const
