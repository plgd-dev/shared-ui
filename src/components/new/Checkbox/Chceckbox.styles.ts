import { css } from '@emotion/react'
import { colors } from '../_utils/colors'
import styled from '@emotion/styled'

export const checkbox = css`
    display: flex;
    align-items: flex-start;
    position: relative;
    overflow: hidden;
`

export const check = styled.span`
    width: 16px;
    height: 16px;
    flex: 0 0 16px;
    border: 1px solid #d7d8da;
    border-radius: 4px;
    cursor: pointer;
    transition: all 0.25s;
    position: relative;
    box-sizing: border-box;

    &:before,
    &:after {
        content: '';
        display: block;
        transition: opacity 0.25s;
        position: absolute;
        top: 50%;
        left: 50%;
        transform: translate(-50%, -50%);
        width: 10px;
        border-radius: 2px;
        background: ${colors.primary};
        opacity: 0;
    }

    &:before {
        height: 2px;
    }

    &:after {
        height: 10px;
    }
`

export const input = css`
    width: 1px;
    height: 1px;
    position: absolute;
    left: -10px;

    &:indeterminate ~ ${check} {
        border-color: ${colors.primary};

        &:before {
            opacity: 1;
        }
    }

    &:checked ~ ${check} {
        border-color: ${colors.primary};

        &:after {
            opacity: 1;
        }
    }
`

export const label = css`
    font-size: 12px;
    color: #81868c;
    padding-left: 8px;
    cursor: pointer;
    line-height: 18px;
    font-family: 'Poppins', sans-serif;
    font-style: normal;
    font-weight: 400;
`

export const error = css`
    border-color: ${colors.error};
`
