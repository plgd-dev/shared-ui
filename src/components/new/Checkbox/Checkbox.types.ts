import { ReactNode, Ref, SyntheticEvent } from 'react'

export type Props = {
    checked?: boolean
    className?: string
    defaultChecked?: boolean
    disabled?: boolean
    error?: boolean
    id?: string
    inputRef?: Ref<any>
    label?: ReactNode
    name: string
    onChange?: (e: SyntheticEvent) => void
    type?: 'checkbox' | 'radio'
    value?: string
}

export const defaultProps: Partial<Props> = {
    type: 'checkbox',
}
