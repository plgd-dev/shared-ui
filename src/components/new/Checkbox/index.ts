export { default } from './Checkbox'
export * from './Checkbox'

export { default as IndeterminateCheckbox } from './IndeterminateCheckbox'
