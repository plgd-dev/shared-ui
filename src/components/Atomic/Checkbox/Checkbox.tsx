import React, { FC, useEffect, useImperativeHandle, useRef } from 'react'
import { defaultProps, Props } from './Checkbox.types'
import * as styles from './Chceckbox.styles'

const Checkbox: FC<Props> = (props) => {
    const { id, className, inputRef, label, checked, disabled, error, indeterminate, topOffset, ...rest } = { ...defaultProps, ...props }
    const Check = styles.check
    const ref = useRef<HTMLInputElement>(null)
    useImperativeHandle(inputRef, () => ref.current!, [ref])

    useEffect(() => {
        if (ref.current) {
            ref.current.indeterminate = indeterminate || false
            if (indeterminate) {
                ref.current.checked = false
            }
        }
    }, [indeterminate, ref])

    return (
        <label className={className} css={styles.checkbox} id={id}>
            <input {...rest} checked={checked} css={styles.input} ref={ref} />
            <Check css={[error && styles.error, topOffset === false && styles.noTopOffset]} />
            {label && <div css={styles.label}>{label}</div>}
        </label>
    )
}

Checkbox.displayName = 'Checkbox'

export default Checkbox
