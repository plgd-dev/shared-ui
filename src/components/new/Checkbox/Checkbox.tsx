import React, { FC } from 'react'
import { defaultProps, Props } from './Checkbox.types'
import * as styles from './Chceckbox.styles'

const Checkbox: FC<Props> = (props) => {
    const { id, className, inputRef, label, checked, disabled, error, ...rest } = { ...defaultProps, ...props }
    const Check = styles.check
    return (
        <label className={className} css={styles.checkbox} id={id}>
            <input {...rest} checked={checked} css={styles.input} ref={inputRef} />
            <Check css={[error && styles.error]} />
            {label && <div css={styles.label}>{label}</div>}
        </label>
    )
}

Checkbox.displayName = 'Checkbox'
Checkbox.defaultProps = defaultProps

export default Checkbox
