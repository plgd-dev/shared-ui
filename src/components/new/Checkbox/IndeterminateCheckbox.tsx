import React, { forwardRef, useRef, useEffect, Ref } from 'react'
import Checkbox from './Checkbox'
import { Props } from './Checkbox.types'

const IndeterminateCheckbox = forwardRef<Ref<any>, Props & { indeterminate: boolean }>((props, ref) => {
    const { indeterminate, ...rest } = props
    const defaultRef = useRef()
    const resolvedRef = ref || defaultRef

    useEffect(() => {
        // @ts-ignore
        resolvedRef.current.indeterminate = indeterminate
    }, [resolvedRef, indeterminate])

    return <Checkbox inputRef={resolvedRef} {...rest} />
})

IndeterminateCheckbox.displayName = 'IndeterminateCheckbox'

export default IndeterminateCheckbox
