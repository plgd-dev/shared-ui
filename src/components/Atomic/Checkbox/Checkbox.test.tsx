import { render } from '@testing-library/react'
import Checkbox from './Checkbox'

describe('<Checkbox>', () => {
    it('render correctly - snapshot', () => {
        const { asFragment } = render(
            <div>
                <Checkbox defaultChecked={false} name='checkbox-01' />
                <Checkbox defaultChecked={true} name='checkbox-02' />
                <Checkbox indeterminate={true} name='checkbox-03' />
                <Checkbox
                    defaultChecked={false}
                    label='Sure, send me insightful product news, and occasional offers by email. I know I can unsubscribe at any time.'
                    name='checkbox-04'
                />
                <Checkbox
                    defaultChecked={true}
                    label='Sure, send me insightful product news, and occasional offers by email. I know I can unsubscribe at any time.'
                    name='checkbox-05'
                />
            </div>
        )

        expect(asFragment()).toMatchSnapshot()
    })
})
