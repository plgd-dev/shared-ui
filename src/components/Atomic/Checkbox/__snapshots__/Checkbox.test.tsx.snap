// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox> render correctly - snapshot 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  position: relative;
  overflow: hidden;
  padding-bottom: 2px;
}

.emotion-1 {
  width: 1px;
  height: 1px;
  position: absolute;
  left: -10px;
}

.emotion-1:indeterminate~.emotion-3 {
  border-color: #2261AE;
}

.emotion-1:indeterminate~.emotion-3:before {
  opacity: 1;
}

.emotion-1:checked~.emotion-3 {
  border-color: #2261AE;
}

.emotion-1:checked~.emotion-3:after {
  opacity: 1;
}

.emotion-2 {
  width: 16px;
  height: 16px;
  -webkit-flex: 0 0 16px;
  -ms-flex: 0 0 16px;
  flex: 0 0 16px;
  border: 1px solid #D7D8DA;
  border-radius: 4px;
  cursor: pointer;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  position: relative;
  box-sizing: border-box;
  top: 2px;
}

.emotion-2:before,
.emotion-2:after {
  content: '';
  display: block;
  -webkit-transition: opacity 0.25s;
  transition: opacity 0.25s;
  position: absolute;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%, -50%);
  -moz-transform: translate(-50%, -50%);
  -ms-transform: translate(-50%, -50%);
  transform: translate(-50%, -50%);
  width: 10px;
  border-radius: 2px;
  background: #2261AE;
  opacity: 0;
}

.emotion-2:before {
  height: 2px;
}

.emotion-2:after {
  height: 10px;
}

.emotion-16 {
  color: #81868C;
  padding-left: 8px;
  cursor: pointer;
}

<div>
    <label
      class="emotion-0"
    >
      <input
        class="emotion-1"
        name="checkbox-01"
        type="checkbox"
      />
      <span
        class="emotion-2 emotion-3"
      />
    </label>
    <label
      class="emotion-0"
    >
      <input
        checked=""
        class="emotion-1"
        name="checkbox-02"
        type="checkbox"
      />
      <span
        class="emotion-2 emotion-3"
      />
    </label>
    <label
      class="emotion-0"
    >
      <input
        class="emotion-1"
        name="checkbox-03"
        type="checkbox"
      />
      <span
        class="emotion-2 emotion-3"
      />
    </label>
    <label
      class="emotion-0"
    >
      <input
        class="emotion-1"
        name="checkbox-04"
        type="checkbox"
      />
      <span
        class="emotion-2 emotion-3"
      />
      <div
        class="emotion-16"
      >
        Sure, send me insightful product news, and occasional offers by email. I know I can unsubscribe at any time.
      </div>
    </label>
    <label
      class="emotion-0"
    >
      <input
        checked=""
        class="emotion-1"
        name="checkbox-05"
        type="checkbox"
      />
      <span
        class="emotion-2 emotion-3"
      />
      <div
        class="emotion-16"
      >
        Sure, send me insightful product news, and occasional offers by email. I know I can unsubscribe at any time.
      </div>
    </label>
  </div>
</DocumentFragment>
`;
